modules/core/src/main/java/org/apache/fluo/core/impl/TransactionImpl.java

```diff
@@ -895,6 +904,22 @@ public int getSize() {
     return size;
   }
 
+  private <V> void addCallback(CompletableFuture<V> cfuture, CommitData cd,
+      OnSuccessInterface onSuccessInterface) {
+    cfuture.thenAcceptAsync(result -> {
+      try {
+        // this method currently throws Exception, so must be caught...
+        onSuccessInterface.onSuccess(cd, result);
+      } catch (Exception e) {
+        cd.commitObserver.failed(e);
+      }
+    }, env.getSharedResources().getAsyncCommitExecutor()).exceptionally(exception -> {
+      // not sure this really needed... since code above catches exception
```

Comment:
Only if `cfuture` completes normally will `thenAcceptAsync` be called.  If `cfuture` completes exceptionally, this is needed to ensure the exception is propagated to the commit observer.  The exception handling code within `thenAcceptAsync` only comes into play after normal completion of `cfuture` and given normal completion of `cfuture` if either `onSuccess` throws an `Error` or `failed` throws an `Error` or `RuntimeException` will `exceptionally` be invoked.

Reply:
This could potentially be done in one step by calling `handleAsync` with a `BiFunction<Result, Throwable, Void>`

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#handleAsync-java.util.function.BiFunction-java.util.concurrent.Executor-

Reply:
I see what you're saying. The catch is necessary in case` OnSuccessInterace.onSuccess()` throws an exception and the `.exceptionally` is necessary in case `cfuture` never completes in the first place.

Reply:
handleAsync would still need need both `failed()` calls correct? something like
```
cfuture.handleAsync((result, exception) -> {
  if(result != null){
    try{
      onSuccessInterface.onSuccess(cd, result);
    } catch (Exception e) {
      cd.commitObserver.failed(e);
    }
  } else {
    cd.commitObserver.failed(exception);
  }
}, env.getSharedResources().getAsyncCommitExecutor());
```
May have left out some `return null`s.